R/pmcmc.R

```diff
@@ -0,0 +1,428 @@
+  # calculate posterior
+  curr_ll <- p_filter_est$log_likelihood
+  curr_lpost <- curr_lprior + curr_ll
+  
+  curr_ss <- p_filter_est$sample_state
+  
+
+  ## combine into a vector to add to the results at the end
+  out_init <- c(curr_pars, 
+                'log_prior' = curr_lprior, 
+                'log_likelihood' = curr_ll, 
+                'log_posterior' = curr_lpost) 
+  
+  ss_init <- curr_ss
+
+  
+  ## initialise output arrays
+  
+  res <- matrix(data = NA, 
+                nrow = n_mcmc, 
+                ncol = length(out_init), 
+                dimnames = list(NULL, names(out_init)))
+  
+  states <- matrix(data = NA,
+                   nrow = n_mcmc, 
+                   ncol = length(ss_init))
+  
+  
+  ## create function to reflect proposal boundaries at pars_min and pars_max
+  # this ensures the proposal is symetrical and we can simplify the MH step
+  
+  reflect_proposal <- function(x, floor, cap) {
+    interval <- cap - floor
+    abs((x + interval - floor) %% (2 * interval) - interval) + floor
+  }
+  
+  
+   # this can easily be adapted to being multivariate normal with covariance to improve mixing
+  propose_jump <- function(pars) {
+    
+    ## proposed jumps are normal with mean pars and sd as input for parameter
+    tmp <- rnorm(n = n_pars, mean = pars, sd = pars_sd)
+    # discretise if necessary
+    tmp[pars_discrete] <- round(tmp[pars_discrete])
+    # reflect proposal if it exceeds upper or lower parameter boundary
+    tmp <- reflect_proposal(x = tmp, 
+                            floor = pars_min, 
+                            cap = pars_max)
+    tmp
+  }
+
+
+  # start progress bar  
+  pb <- txtProgressBar(min = 0, max = n_mcmc, style = 3)
+  
+  # main pmcmc loop
+  for(iter in seq_len(n_mcmc)) {
+    
+    setTxtProgressBar(pb, iter)
+    
+    # propose new parameters
+    prop_pars <- propose_jump(curr_pars)
+    
+    ## calculate proposed prior / lhood / posterior 
+    prop_lprior <- calc_lprior(pars = prop_pars)
+    p_filter_est <- calc_ll(pars = curr_pars)
+    prop_ll <- p_filter_est$log_likelihood
+    prop_ss <- p_filter_est$sample_state
+    prop_lpost <- prop_lprior + prop_ll
+    
+    
+    # calculate probability of acceptance
+    accept_prob <- exp(prop_lpost - curr_lpost)
+
+    
+    if(runif(1) < accept_prob) { # MH step
+      # update parameters and calculated likelihoods
+      curr_pars <- prop_pars
+      curr_lprior <- prop_lprior
+      curr_ll <- prop_ll
+      curr_lpost <- prop_lpost
+      curr_ss <- prop_ss
+    }
+    
+    # record results
+    res[iter, ] <- c(curr_pars, 
+                     'log_prior' = curr_lprior, 
+                     'log_likelihood' = curr_ll, 
+                     'log_posterior' = curr_lpost) 
+    states[iter, ] <- curr_ss
+
+  }
+  # end progress bar
+  close(pb)
+  
+  
+  # include starting conditions in output
+  res <- rbind(out_init, res, deparse.level = 0)
+  res <- as.data.frame(res)
+  
+  tmp <- coda::as.mcmc(res)
```

Comment:
Change `tmp` to `coda_res` or similar

Reply:
Actually just remove this, if it can be added above

Reply:
done and pushed